### PR TITLE
iox-#2067 Support move-only types in the LockFreeQueue

### LINKED ISSUE #2067: LockFreeQueue fails to support move-only types
## Required information

**Operating system:**
Ubuntu 20.04 LTS

**Compiler version:**
GCC 9.4.0

**Eclipse iceoryx version:**
`master` branch

**Observed result or behaviour:**
`iox::concurrent::LockFreeQueue` fails to compile with move-only types.

**Expected result or behaviour:**
`iox::concurrent::LockFreeQueue` should support move-only types.

**Conditions where it occurred / Performed steps:**
Make the `Integer` struct in [test_concurrent_lockfree_queue.cpp](https://github.com/eclipse-iceoryx/iceoryx/blob/69125796fe2df4645888a41ebf25cd7d105e63ab/iceoryx_hoofs/test/moduletests/test_concurrent_lockfree_queue.cpp#L34) move-only and observe when running `bazel test //iceoryx_hoofs/test:hoofs_moduletests`:
```
In file included from iceoryx_hoofs/include/iceoryx_hoofs/concurrent/lockfree_queue.hpp:131,
                 from iceoryx_hoofs/test/moduletests/test_concurrent_lockfree_queue.cpp:20:
iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/lockfree_queue.inl: In instantiation of 'void iox::concurrent::LockFreeQueue<ElementType, Capacity>::writeBufferAt(const uint64_t&, T&&) [with T = {anonymous}::Integer&; ElementType = {anonymous}::Integer; long unsigned int Capacity = 10; uint64_t = long unsigned int]':
iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/resizeable_lockfree_queue.inl:216:24:   required from 'iox::optional<OptionalBaseType> iox::concurrent::ResizeableLockFreeQueue<ElementType, Capacity>::pushImpl(T&&) [with T = {anonymous}::Integer; ElementType = {anonymous}::Integer; long unsigned int MaxCapacity = 10]'
iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/resizeable_lockfree_queue.inl:190:53:   required from 'iox::optional<OptionalBaseType> iox::concurrent::ResizeableLockFreeQueue<ElementType, Capacity>::push(ElementType&&) [with ElementType = {anonymous}::Integer; long unsigned int MaxCapacity = 10]'
iceoryx_hoofs/test/moduletests/test_concurrent_lockfree_queue.cpp:313:9:   required from 'void {anonymous}::LockFreeQueueTest_pushInsertsInFifoOrder_Test<gtest_TypeParam_>::TestBody() [with gtest_TypeParam_ = {anonymous}::Config<iox::concurrent::ResizeableLockFreeQueue, {anonymous}::Integer, 10, 9>]'
iceoryx_hoofs/test/moduletests/test_concurrent_lockfree_queue.cpp:301:1:   required from here
iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/lockfree_queue.inl:173:5: error: use of deleted function '{anonymous}::Integer::Integer(const {anonymous}::Integer&)'
  173 |     new (elementPtr) ElementType(std::forward<T>(value)); // move ctor invoked when available, copy ctor otherwise
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
iceoryx_hoofs/test/moduletests/test_concurrent_lockfree_queue.cpp:44:5: note: declared here
   44 |     Integer(const Integer&) = delete;
      |     ^~~~~~~
```

### PR DESCRIPTION
**Pull request by gpalmer-latai**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #2067

### CODE REVIEW

File: iceoryx_hoofs/test/moduletests/test_concurrent_lockfree_queue.cpp [hunk 1]
```diff
@@ -39,6 +40,14 @@ struct Integer
...
     }
 
+    // Delete copy constructor to ensure that the queue supports move-only types.
+    Integer(const Integer&) = delete;
+    Integer& operator=(const Integer&) = delete;
+    Integer(Integer&&) = default;
+    Integer& operator=(Integer&&) = default;
+
+    ~Integer() = default;
+
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_concurrent_lockfree_queue.cpp:50]: Could you revert this and create a `IntegerMoveOnly` and `IntegerCopyOnly` struct (maybe inherit from `integer`) and add two tests for each of the two structs for `push` and `tryPush`, e.g.
```cpp
TEST(LockFreeQueueMoveTest, PushPopWithMoveOnlyTypeWorks)
{
    ::testing::Test::RecordProperty("TEST_ID", "dc85a403-8265-4853-8360-73ea66668f24");

    iox::concurrent::LockFreeQueue<IntegerMoveOnly, 10> sut;
    const int EXPECTED_VALUE{42};
    sut.push({EXPECTED_VALUE});
    sut.pop().and_then([&](const auto& value) { EXPECT_THAT(value, Eq(EXPECTED_VALUE)); }).or_else([&] {
        GTEST_FAIL() << "Expected '" << EXPECTED_VALUE << "' but got nothing!";
    });
}
```
Could be done before `LockFreeQueueTest` class definition.

The other tests can remain as they are since they test the internal logic and if `push`, `try_push` and `pop` work with move only types everything will work.

**Review by elBoberido — COMMENTED** [review]
Good catch. The fix is fine but could you add some tests for move only and copy only types please :)

File: doc/website/release-notes/iceoryx-unreleased.md [hunk 2]
```diff
@@ -54,6 +54,7 @@
 
 **Bugfixes:**
 
+- LockFreeQueue fails to support move-only types [\#2067](https://github.com/eclipse-iceoryx/iceoryx/issues/2067)
```
[the remark this answers is not in the document]
Comment (gpalmer-latai) [reply doc/website/release-notes/iceoryx-unreleased.md:57]: Good to know. I may as well change now along with the other changes :)

File: iceoryx_hoofs/test/moduletests/test_concurrent_lockfree_queue.cpp (hunk 1 above)
Comment (gpalmer-latai) [reply iceoryx_hoofs/test/moduletests/test_concurrent_lockfree_queue.cpp:50]: I certainly can and will refactor to add additional distinct `MoveOnly` and `CopyOnly` versions. 

I wonder however if it would not be more beneficial to simply add those types to the existing type list so they get included in all of the existing tests. Is the idea here to preserve the integrity of the existing tests?

**Comment by gpalmer-latai** [comment]
> Could you revert this and create a IntegerMoveOnly and IntegerCopyOnly struct

Amazing, this isn't even an edge case I had considered.

```
iceoryx_hoofs/test/moduletests/test_concurrent_lockfree_queue.cpp:95:20:   required from here
iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl:281:5: error: use of deleted function '{anonymous}::CopyOnlyInteger::CopyOnlyInteger({anonymous}::CopyOnlyInteger&&)'
  281 |     new (&m_data) T(std::forward<Targs>(args)...);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
iceoryx_hoofs/test/moduletests/test_concurrent_lockfree_queue.cpp:65:5: note: declared here
   65 |     CopyOnlyInteger(CopyOnlyInteger&&) = delete;
```

**Comment by gpalmer-latai** [comment]
Alright, I didn't think too hard about this this morning but upon a bit further research a realized/remembered that "copy only types" aren't really a thing. It's technically possible to define a type that can only be assigned to / constructed from an lvalue but not an rvalue, but I'm not aware of any practical reason to do so, and the standard library algorithms don't bother supporting such constructs either. In fact `std::copyable` will fail on a type with a deleted move constructor.

So I'm going to just add the `MoveOnlyInteger` and associated unit tests.

File: iceoryx_hoofs/test/moduletests/test_concurrent_lockfree_queue.cpp (hunk 1 above)
Comment (elBoberido) [reply iceoryx_hoofs/test/moduletests/test_concurrent_lockfree_queue.cpp:50]: You can of course add them to the type list.  At a first glance the setup looked a bit complicated and I was not sure how easy it would be to add tests with a different element type. It's up to you :)

Comment (gpalmer-latai) [reply iceoryx_hoofs/test/moduletests/test_concurrent_lockfree_queue.cpp]: It turned out to be pretty easy. Just define the new type and add it to a bunch of templates. I confirmed by undoing the `std::forward` fix briefly to observe the failure recurring.

**Comment by gpalmer-latai** [comment]
I've updated the tests to simply define a distince `MoveOnlyInteger` type and add it to the templates used by the typed test suite. I confirmed by undoing the bugfix that this does indeed cover the move-only case.

**Review by elBoberido — APPROVED** [review]
Looks good. Thanks very much for your contribution

**Review by FerdinandSpitzschnueffler — APPROVED** [review]
Thanks for your contribution!

**Comment by elBoberido** [comment]
@gpalmer-latai the action is kind of broken. See here https://github.com/vmactions/freebsd-vm/issues/74. I'll trigger it another two or three times and then merge it anyway. These changes will not break FreeBSD when it runs fine on all the other platforms. FreeBSD usually only breaks when we have POSIX related changes.

If we get an [approval for Cirrus CI](https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/3905) the FreeBSD build will become the fastest and most stable one.

Btw. on Thursday is the iceoryx developer meetup in case you want to get in touch with us and tell us about your experience with and wishes for iceoryx :)

**Comment by gpalmer-latai** [comment]
> Btw. on Thursday is the iceoryx developer meetup in case you want to get in touch with us and tell us about your experience with and wishes for iceoryx :)

@elBoberido Is it at 17:00 CET with the same zoom link as the October one? It isn't on the calendar as stated in https://github.com/eclipse-iceoryx/iceoryx/wiki/Developer-meetup

**Comment by elBoberido** [comment]
@gpalmer-latai yes, it is 17:00 CET but we want to try a more privacy friendly solution than zoom. I have to add a new wiki page but this will be the link https://app.element.io/jitsi.html#conferenceDomain=meet.element.io&conferenceId=JitsiVtfrqukadefbqiqfxryxabai&userId=&roomId=!AooDAAwkyNWwkMElpt%3Agitter.im&roomName=eclipse%2Ficeoryx&startWithAudioMuted=true&startWithVideoMuted=true&language=en

@mossmaurice could you please check why the meeting is not on the ROS calendar? I had the impression it would have been added again.

**Comment by gpalmer-latai** [comment]
Cool. I'll add a calendar event for 12:00 PM EDT tomorrow (Germany always changes their clock one or two weeks earlier than the US) 😉
